Fix test_iocp.py and test_tcp.py on Windows with Python 3

rodrigc opened the pull request:
See:
https://twistedmatrix.com/trac/ticket/8631

rodrigc committed:
Add topfile

rodrigc committed:
Import twisted.python.compat.unicode for Python 2/3 compatibility

rodrigc committed:
Switch to using a dataBuffer of bytes, not str

rodrigc committed:
replace buffer() with memoryview()

buffer() is gone in Python 3

rodrigc committed:
Use socket.error.errno directly instead of unpacking socket.error.args

On Python 3, socket.error.args unpacks to a larger tuple than on
Python 2, so this code fails on Python 3

rodrigc committed:
Switch to using the Win32 API WSAAddressToStringW() which uses wchar_t

WSAAddressToStringA() which uses ASCII is deprecated

rodrigc committed:
Switch to using the Win32 API WSAAddressToStringW() which uses wchar_t

WSAAddressToStringA() which uses ASCII is deprecated.
Eliminate inet_ntoa() and inet_addr() which do
not support wchar_t.

rodrigc committed:
On Python 2, convert host to unicode before passing to PyUnicode_AsWideChar()

This is not needed on Python 3, since Python 3 strings are unicode.
This fixes the test_tcp test on Python 2 after the last round of
changes.

rodrigc committed:
Regenerate with Cython 0.24.1

rodrigc commented:
@opalmer can you review this?  This patch is needed to get the Twisted IOCPReactor to work on Python 3.

glyph commented:
Hrm. Someone turned codecov status reporting back on? :)  It looks like it's missing the coverage reports from the py3 builders: https://codecov.io/gh/twisted/twisted/compare/3e488bca70159bd8d7f5a62dc55e0ab9879f229b...e5b0d78199284b6da26beec8c833bfd5a1ee782d?src=pr#747769737465642F696E7465726E65742F696F637072656163746F722F7463702E7079-64

glyph commented on twisted/internet/iocpreactor/tcp.py:
```diff
@@ -59,7 +60,13 @@ def getHandle(self):
         # XXX: some day, we'll have protocols that can handle raw buffers
-        self.protocol.dataReceived(str(rbuffer))
+        if not isinstance(rbuffer, bytes):
+            if _PY3:
```
Anywhere we use `_PY3` it's a potential future maintenance problem.  We should strive to always use the relevant high-level API from twisted.python.compat.  In this case though I don't know what type the `rbuffer` argument might be, so it's hard to be sure what exactly this should be.  I'm like 90% sure that `"utf-8"` is the wrong thing to say below, but without being surer of what this is I don't know what else to suggest.

To address this, as per [policy](https://twistedmatrix.com/trac/wiki/ReviewProcess#Authors:Thingsyourbranchorpatchmustcontain), all new or changed methods must have 100% docstring coverage, so if you could add a docstring here that says what types `rbuffer` is allowed to be.

rodrigc replied:
There was a Windows Python 3 buildbot, but adiroiban disabled it and deleted it.
Right now the tox.ini file records coverage, and the appveyor job which runs tox uploads the coverage report to codecov.io.  However, the appveyor job is only doing Windows Python 2.

We can't enable a Windows Python 3 buildbot / appveyor just yet, because some of the jobs hang, but this patch is one big step in getting us closer.

rodrigc replied:
This dataReceived() call is nasty.  I went through an experiment of changing this
to not convert the argument at all, and just passthru the rbuffer argument to the next layer.  However, during the full test run, some of the tests failed, because they hit assertions elsewhere in the code where data was detected as type(unicode) (even on Python 2).

Some of those other protocols that pass unicode to write() methods need to be cleaned up,
but that is way out of the scope of this PR.

glyph commented on twisted/internet/iocpreactor/iocpsupport/iocpsupport.pyx:
```diff
@@ -214,55 +218,89 @@ def makesockaddr(object buff):
     cdef unsigned short port
-    cdef unsigned long res
-    cdef char *hoststr
+    cdef WCHAR hostStr[256]
```
In Python code, I'm usually against comments, but in C (or C-esque, I guess?) code it seems important to call stuff out.  256 is an appropriate value here because that's a little longer than the longest possible domain name: http://stackoverflow.com/questions/32290167/what-is-the-maximum-length-of-a-dns-name#32294443 - right?  Stack allocation always makes me nervous :).  (It looks like `hoststr` used to just be leaked though?)

rodrigc replied:
That's right 256 is a bit longer than the longest domain name.
It's better not to leak, in case something goes wrong.

rodrigc committed:
Fix whitespace

rodrigc committed:
Add comment

rodrigc committed:
Add docstring for dataReceived().

rodrigc committed:
dataReceived() should only accept bytes or bytearray

rodrigc committed:
On Python 2, convert the hostname obtained from _makesockaddr() to str from unicode

rodrigc committed:
Rename bytes variable to data, to avoid confusion with Python bytes type

rodrigc committed:
Use bytes instead of str for Python 3

opalmer commented:
@rodrigc sorry for the delay, I've been traveling.  Changes on the Windows side seem ok to me, @glyph will probably want to take a final look at the new changes since this last review.  Other than that, thanks for looking into this!

rodrigc committed:
Merge branch 'trunk' into 8631-rodrigc-iocp-py3

rodrigc committed:
Rename variable to indicate number of bytes, and not data buffer.

getEvent() calls the Windows API GetQueuedCompletionStatus()
which returns the number of bytes read or written, not the actual
bytes buffer.

rodrigc committed:
Pydoctor cannot handle L{bytearray}

rodrigc committed:
Move comment

rodrigc committed:
Merge branch 'trunk' into 8631-rodrigc-iocp-py3

rodrigc committed:
Try L{bytearray}

rodrigc committed:
Use byte string from Python 3

rodrigc committed:
Enable Appveyor build on Python 3.5

rodrigc committed:
Use the default value of TRIAL_REPORTER=verbose in tox.ini

Appveyor does not seem to have the 4MB build log restriction that
Travis has.

rodrigc committed:
Merge branch 'trunk' into 8631-rodrigc-iocp-py3

rodrigc committed:
Add topfile for #8678

markrwilliams commented on twisted/internet/iocpreactor/abstract.py:
```diff
@@ -71,8 +72,8 @@ def _dispatchData(self):
-            self.dataReceived(buffer(self._readBuffers[full_buffers],
-                                     0, remainder))
+            remainderView = memoryview(self._readBuffers[full_buffers])
+            self.dataReceived(remainderView[0:remainder].tobytes())
```
calling `tobytes()` on the `memoryview` will copy the underlying array to the new `bytes` object, defeating the purpose of using a `memoryview` in the first place.

if this code used to accept `buffer`, though, it should also accept `memoryview`, since `memoryview`'s API is a superset of `buffer`'s.  so it should be enough to just change this line to:

``` python
self.dataReceived(remainderView[0:remainder])
```

(you can remove the 0 too, if that reads better)

here's my citation for the copying claim.  the code for `tobytes()` is [here](https://github.com/python/cpython/blob/2.7/Objects/memoryobject.c#L405-L418); if `PyBuffer_Release` isn't suspicious enough, you can see from the source for [`PyBuffer_ToContiguous`](https://github.com/python/cpython/blob/2.7/Objects/abstract.c#L482-L531) that it [copies](https://github.com/python/cpython/blob/2.7/Objects/abstract.c#L496) the backing array, [even if it's not contiguous](https://github.com/python/cpython/blob/2.7/Objects/abstract.c#L523-L528).

markrwilliams commented on twisted/internet/iocpreactor/abstract.py:
```diff
@@ -236,8 +237,9 @@ def doWrite(self):
-            self.dataBuffer = (buffer(self.dataBuffer, self.offset) +
-                               "".join(self._tempDataBuffer))
+            writeView = memoryview(self.dataBuffer)
+            self.dataBuffer = (writeView[self.offset:].tobytes() +
```
same comment on `memoryview.tobytes()` as before

rodrigc replied:
If I remove memoryview.tobytes() from the places that you recommended, then
when I run the tests, some of them start hanging, such as this one:

```
twisted.internet.test.test_tcp
  AbortConnectionTests_IOCPReactorTests
    test_clientAbortsConnectionThenLosesConnection ...                  [ERROR]
    test_clientAbortsConnectionTwice ...                                [ERROR]
    test_dataReceivedAbort ...                                                [ERROR]
```

If I dig into it some more and run the test outside of trial with:

```
python -m unittest twisted.internet.test.test_tcp
```

I get:

```
Unhandled Error
Traceback (most recent call last):
  File "twisted\internet\test\connectionmixins.py", line 178, in runProtocolsWit
hReactor
    reactorBuilder.runReactor(reactor)
  File "twisted\internet\test\reactormixins.py", line 314, in runReactor
    reactor.run()
  File "twisted\internet\base.py", line 1195, in run
    self.mainLoop()
  File "twisted\internet\base.py", line 1204, in mainLoop
    self.runUntilCurrent()
--- <exception caught here> ---
  File "twisted\internet\base.py", line 825, in runUntilCurrent
    call.func(*call.args, **call.kw)
  File "twisted\internet\iocpreactor\abstract.py", line 188, in _resumeWriting
    self.doWrite()
  File "twisted\internet\iocpreactor\abstract.py", line 242, in doWrite
    b"".join(self._tempDataBuffer))
exceptions.TypeError: unsupported operand type(s) for +: 'memoryview' and 'str'
```

markrwilliams commented on twisted/internet/iocpreactor/abstract.py:
```diff
@@ -236,8 +237,9 @@ def doWrite(self):
-                               "".join(self._tempDataBuffer))
+            writeView = memoryview(self.dataBuffer)
+            self.dataBuffer = (writeView[self.offset:].tobytes() +
+                               b"".join(self._tempDataBuffer))
```
of course, you can't do `memoryview + bytes`, so your choices are either to drop `memoryview` entirely or keep `tobytes()`.

i think you should drop the `memoryview` entirely.  as i said in a previous comment, using it and then calling `tobytes()` copies the backing array, apparently negating speed and memory savings.  it might seem that slicing into a `memoryview` with a large `self.offset` will only copy the last part of the backing array, but this is actually how `bytes` and `str` slicing already works in cpython.  because strings are immutable, slices can be shared.

tl;dr: just do:

``` python
self.dataBuffer = self.dataBuffer[self.offset:] + b"".join(self._tempDataBuffer)
```

rodrigc committed:
Remove two uses of memoryview

glyph committed:
Merge branch 'trunk' into 8631-rodrigc-iocp-py3

glyph commented:
Just re-running that one Windows build in the hopes that it will pass this time (come _on_ `test_manyProcesses`... I am starting to think we should just delete that test)

glyph commented:
I'm going to finish waiting for this one run in the hopes of an all-green status but I'm seeing the same failure on unrelated branches, so if I just get it again I'm going to land regardless.

glyph commented:
Wow. We have to wait quite a while for a build result.

glyph commented:
And it's off! https://buildbot.twistedmatrix.com/builders/windows7-64-py2.7/builds/1386

glyph commented:
GREEEEEEEEEEEEEEEEEEEEEEEEN

rodrigc commented:
@glyph @runciter Thanks for reviewing and merging this.  I put some nontrivial effort into working on this and debugging.  This pushed the limits of my Python knowledge, as I had to learn
about debugging Twisted reactor failures, Cython, Python C API, buffer API, memoryview API, Win32 API, and Python 3 on Windows.

This is probably the most challenging Python bug I have ever worked on!

glyph commented:
Thanks for getting it landed, @rodrigc ! As you have discovered, this is some of the most obscure and challenging code to work on in Twisted and it is _awesome_ that we have another person who has onboarded into it :)
